### PR TITLE
ARROW-3278: [Python] Retrieve StructType's and StructArray's field by name

### PR DESCRIPTION
**Pull request by kszucs**

### CODE REVIEW

File: python/pyarrow/array.pxi [hunk 1]
```diff
@@ -1146,6 +1146,28 @@ cdef class StructArray(Array):
             CStructArray* sarr = <CStructArray*> self.ap
         return pyarrow_wrap_array(sarr.field(ix))
 
+    def field_by_name(self, str name):
```
Comment (pitrou) [inline python/pyarrow/array.pxi]: Can't this just be a `.field()` overload? `field_by_name` is a bit verbose.

Comment (kszucs) [reply python/pyarrow/array.pxi]: @wesm requested `field_by_name` in the issue https://issues.apache.org/jira/browse/ARROW-3278

Comment (wesm) [reply python/pyarrow/array.pxi]: I'd be fine with overloading `field`. If we do that, we should also do this in `Schema.field` (where there is also `field_by_name`

Comment (kszucs) [reply python/pyarrow/array.pxi]: Ok

Comment (pitrou) [reply python/pyarrow/array.pxi]: Either that, or raise an exception. We can revise later, IMO.

Comment (kszucs) [reply python/pyarrow/array.pxi]: Ok, I'll leave it as is. I'd consider a warning too instead of an exception, should I create a jira?

Comment (pitrou) [reply python/pyarrow/array.pxi]: That sounds fine to me.

**Comment by kszucs** [comment]
Created two jira tickets:
- https://issues.apache.org/jira/browse/ARROW-3532
- https://issues.apache.org/jira/browse/ARROW-3531

File: python/pyarrow/types.pxi [hunk 2]
```diff
@@ -207,6 +211,19 @@ cdef class StructType(DataType):
 
     cdef void init(self, const shared_ptr[CDataType]& type):
         DataType.init(self, type)
+        self.struct_type = <const CStructType*> type.get()
+
+    cdef Field child_by_name(self, name):
```
Comment (wesm) [inline python/pyarrow/types.pxi]: Probably should call this `field_by_name` too

File: python/pyarrow/types.pxi [hunk 3]
```diff
@@ -207,6 +211,19 @@ cdef class StructType(DataType):
...
+        """
+        Access a child field by its name rather than the column index.
+        """
+        cdef shared_ptr[CField] field
+
+        field = self.struct_type.GetChildByName(tobytes(name))
```
Comment (wesm) [inline python/pyarrow/types.pxi]: I think we should rename this to `GetFieldByName` and `field` in `arrow::StructType`. Doesn't need to be in this patch

**Review by wesm — APPROVED** [review]
Seeing this, I'm unsure again about the child/field terminology. In the case of StructType, I would prefer to refer to things as "fields". But for a general nested type, "child" is OK

**Comment by wesm** [comment]
I opened https://issues.apache.org/jira/browse/ARROW-3545 about normalizing terminology as it makes sense

**Pull request closed without merging**